modules/sharedIdSystem.js

```diff
@@ -0,0 +1,210 @@
+ */
+function existingIdCallback(storedId, callback) {
+  return {
+    success: function (responseBody) {
+      utils.logInfo('SharedId: id to be synced: ' + storedId.id);
+      if (responseBody) {
+        try {
+          let responseObj = JSON.parse(responseBody);
+          storedId = constructCookieValue(responseObj.sharedId, false);
+          utils.logInfo('SharedId: Older SharedId: ' + storedId.id);
+        } catch (error) {
+          utils.logError(error);
+        }
+      }
+      callback(storedId);
+    },
+    error: function () {
+      utils.logInfo('SharedId: Sync error for id : ' + storedId.id);
+      callback(storedId);
+    }
+  }
+}
+
+/**
+ * Encode the id
+ * @param value
+ * @returns {string|*}
+ */
+function encodeId(value) {
+  const result = {};
+  const sharedId = (value && typeof value['id'] === 'string') ? value['id'] : undefined;
+  if (sharedId == OPT_OUT_VALUE) {
+    return undefined;
+  }
+  if (sharedId) {
+    const bidIds = {
+      id: sharedId,
+    }
+    const ns = (value && typeof value['ns'] === 'boolean') ? value['ns'] : undefined;
+    if (ns == undefined) {
+      bidIds.third = sharedId;
+    }
+    result.sharedid = bidIds;
+    utils.logInfo('SharedId: Decoded value ' + JSON.stringify(result));
+    return result;
+  }
+  return sharedId;
+}
+
+/** @type {Submodule} */
+export const sharedIdSubmodule = {
+  /**
+   * used to link submodule with config
+   * @type {string}
+   */
+  name: MODULE_NAME,
+
+  /**
+   * decode the stored id value for passing to bid requests
+   * @function
+   * @param {string} value
+   * @returns {{sharedid:{ id: string, third:string}} or undefined if value doesn't exists
+   */
+  decode(value) {
+    return (value) ? encodeId(value) : undefined;
+  },
+
+  /**
+   * performs action to obtain id and return a value.
+   * @function
+   * @param {SubmoduleParams} [configParams]
+   * @returns {sharedId}
+   */
+  getId(configParams) {
+    const resp = function (callback) {
+      utils.logInfo('SharedId: Sharedid doesnt exists, new cookie creation');
+      ajax(ID_SVC, idGenerationCallback(callback), undefined, {method: 'GET', withCredentials: true});
+    };
+    return {callback: resp};
+  },
+
+  /**
+   * performs actions even if the id exists and returns a value
+   * @param configParams
+   * @param storedId
+   * @returns {{callback: *}}
+   */
+  extendId(configParams, storedId) {
+    utils.logInfo('SharedId: Existing shared id ' + storedId.id);
+    const resp = function (callback) {
+      const needSync = isIdSynced(configParams, storedId);
+      if (needSync) {
+        utils.logInfo('SharedId: Existing shared id ' + storedId + ' is not synced');
+        const sharedIdPayload = {};
+        sharedIdPayload.sharedId = storedId.id;
+        const payloadString = JSON.stringify(sharedIdPayload);
+        ajax(ID_SVC, existingIdCallback(storedId, callback), payloadString, {method: 'POST', withCredentials: true});
```

Comment:
is it intentional that this is a POST call while the `getId` method uses a GET?

Reply:
Yes this is intentional as we want to send the existing id value

Reply:
Ok...you could still use a GET to pass it as a querystring param, but obviously no problem to do it as a POST, I just wanted to make sure it was intentional.